File: core/src/main/java/org/selyu/smp/core/command/resolver/ProfileContextResolver.java
package org.selyu.smp.core.command.resolver;

import co.aikar.commands.BukkitCommandExecutionContext;
import co.aikar.commands.InvalidCommandArgument;
import co.aikar.commands.contexts.ContextResolver;
import org.selyu.smp.core.Core;
import org.selyu.smp.core.data.Repository;
import org.selyu.smp.core.manager.ProfileManager;
import org.selyu.smp.core.profile.Profile;

import java.util.Optional;

public final class ProfileContextResolver implements ContextResolver<Profile, BukkitCommandExecutionContext> {
    private final ProfileManager profileManager = Core.getInstance().getProfileManager();
    private final Repository repository = Core.getInstance().getRepository();

    @Override
    public Profile getContext(BukkitCommandExecutionContext bukkitCommandExecutionContext) throws InvalidCommandArgument {
        String username = bukkitCommandExecutionContext.popFirstArg();
        Optional<Profile> profileInCache = profileManager.getByUsername(username);
        if (profileInCache.isPresent()) {
            return profileInCache.get();
        } else {
            Optional<Profile> profileInRepository = repository.getProfileStore().getByUsername(username).join();
            return profileInRepository.orElseThrow(() -> new InvalidCommandArgument(String.format("Invalid profile! Has %s logged on the server?", username)));
        }
    }
}

File: core/src/main/kotlin/org/selyu/smp/core/command/resolver/ProfileContextResolver.kt
package org.selyu.smp.core.command.resolver

import co.aikar.commands.BukkitCommandExecutionContext
import co.aikar.commands.InvalidCommandArgument
import co.aikar.commands.contexts.ContextResolver
import com.ea.async.Async.await
import org.selyu.smp.core.Core
import org.selyu.smp.core.profile.Profile

class ProfileContextResolver : ContextResolver<Profile, BukkitCommandExecutionContext> {
    private val profileManager = Core.instance.profileManager
    private val repository = Core.instance.repository

    override fun getContext(c: BukkitCommandExecutionContext): Profile {
        val username = c.popFirstArg()
        val fromCache = profileManager.getProfileByUsername(username)
        return if (fromCache.isPresent) {
            fromCache.get()
        } else {
            val fromStore = await(repository.profileStore.getByUsername(username))
            if (fromStore.isPresent) {
                fromStore.get()
            } else {
                throw InvalidCommandArgument("Invalid profile! Has '$username' logged on before?")
            }
        }
    }
}